design/auth.md

```diff
@@ -0,0 +1,270 @@
+# Auth Service
+
+## Context
+
+This is a rewrite of how we do authentication and authorization in the Screwdriver world.  The purpose of this is to harden our security as well as enable us to support private repositories.  Much of what is being proposed here is based on the [JWT Spec](https://tools.ietf.org/html/draft-ietf-oauth-json-web-token-32) and Docker Hub's [implementation](https://docs.docker.com/registry/spec/auth/jwt/).
+
+## New Service (Auth)
+
+We should add a new service, called `auth`, that owns authentication and authorization.  This will be the entry point for users to authenticate via OAuth or Access Tokens as well authorize Bearer Tokens for API access.
+
+### Token Types
+
+ - *Refresh* - 12h signed JWT for authenticating a user for the Auth service.
+ - *Bearer* - 5m signed JWT for authorizing specific operations against the Screwdriver API.
+
+### Login Types
+
+Tokens will be given out for logged-in users, anonymous users, and builds.  Subject will be based on the type:
+
+ - Logged-in User - `user:<YOUR-NAME-HERE>`
+ - Anonymous User - `anon:<RANDOM-ID-HERE>`
+ - Build - `build:<BUILD-ID-HERE>`
+
+## Flow Changes
+
+![public-repo](./diagrams/auth-repo-public.png)
+![private-repo](./diagrams/auth-repo-private.png)
+![private-mode](./diagrams/auth-private.png)
+![login-anon](./diagrams/auth-login-anon.png)
+![login-scm](./diagrams/auth-login-scm.png)
+
+## API Changes
+
+When accessing the Screwdriver API, API should validate that the user has sent a Bearer token valid for that API resource.  If not, it should reject with `401 Unauthorized` with a RFC7235 valid `WWW-Authenticate` header indicating how to get the appropriate Bearer token.
+
+### Search
+
+Pipeline List API will _only_ return "publicly accessible" pipelines.  That means "private" repositories will not show up in search.  Additionally, if Screwdriver is in "Required Login" mode, then that API will still return "publicly accessible" repositories, but require a valid user Bearer token.
+
+### Resources
+
+The resources (or scope) that we care about are `pipelines`, their `jobs`, and their `builds`.  There are two types of permission, `read` and `write`.
```

Comment:
Missing here is a discussion on template ownership and permissions. Forward thinking, we may also want to do something with shared steps as we do with templates.